### PR TITLE
Fix driver verifier failure from zero size allocation

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Signed-off-by: Alan Jowett <alan.jowett@microsoft.com>

## Description

Zero-sized ExAllocatePool is not permitted. Add assert to catch it and fix the case in native program load.

## Testing

CI/CD with driver verifer.

## Documentation

No.

Resolves: #1115

### CODE REVIEW

**Review by saxena-anurag — APPROVED** [review]

**Review by dthaler — APPROVED** [review]

**Review by shankarseal — APPROVED** [review]

**Comment by Alan-Jowett** [comment]
This is hitting an assert during fuzzing and needs more investigation.

File: tools/bpf2c/bpf_code_generator.cpp [hunk 1]
```diff
@@ -941,39 +944,46 @@ bpf_code_generator::emit_c_code(std::ostream& output_stream)
...
     }
-    output_stream << "};" << std::endl;
-    output_stream << "#pragma data_seg(pop)" << std::endl;
-    output_stream << std::endl;
     output_stream << "static void" << std::endl
                   << "_get_programs(_Outptr_result_buffer_(*count) program_entry_t** programs, _Out_ size_t* count)"
                   << std::endl;
     output_stream << "{" << std::endl;
-    output_stream << INDENT "*programs = _programs;" << std::endl;
+    if (sections.size() != 0) {
```
Comment (saxena-anurag) [inline tools/bpf2c/bpf_code_generator.cpp:982]: same as above.

File: tools/bpf2c/bpf_code_generator.cpp [hunk 2]
```diff
@@ -941,39 +944,46 @@ bpf_code_generator::emit_c_code(std::ostream& output_stream)
...
-        output_stream << INDENT INDENT "\"" << program.program_name.c_str() << "\"," << std::endl;
-        output_stream << INDENT INDENT << map_array_name.c_str() << "," << std::endl;
-        output_stream << INDENT INDENT << program.referenced_map_indices.size() << "," << std::endl;
-        output_stream << INDENT INDENT << helper_array_name.c_str() << "," << std::endl;
-        output_stream << INDENT INDENT << program.helper_functions.size() << "," << std::endl;
-        output_stream << INDENT INDENT << program.output.size() << "," << std::endl;
-        output_stream << INDENT INDENT "&" << sanitize_name(program_type_guid_name) << "," << std::endl;
-        output_stream << INDENT INDENT "&" << sanitize_name(attach_type_guid_name) << "," << std::endl;
-        output_stream << INDENT "}," << std::endl;
+    if (sections.size() != 0) {
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply tools/bpf2c/bpf_code_generator.cpp:947]: The verifier permits BPF programs with no code segments, so why should bpf2c block it?

File: tools/bpf2c/bpf_code_generator.cpp (hunk 1 above)
Comment (Alan-Jowett) [reply tools/bpf2c/bpf_code_generator.cpp:982]: The verifier permits BPF programs with no code segments, so why should bpf2c block it?

File: tools/bpf2c/bpf_code_generator.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (saxena-anurag) [reply tools/bpf2c/bpf_code_generator.cpp:947]: hmm, actually it is not the verifier which allows it. It is our loader which allows it (ebpfapi calls into verifier for each program it finds in the ELF file). Though I agree we can allow loading such native drivers. 

But from a real scenario, creating a native driver just for map creation is over-kill.

Also, native module code in ebpfcore today rejects such module if program count it 0 (even maps defined in the native driver will not be created). So we may as well stop generating code for such ELF files in bpf2c also.

File: tests/sample/empty.c [hunk 3]
```diff
@@ -0,0 +1,28 @@
+// Copyright (c) Microsoft Corporation
+// SPDX-License-Identifier: MIT
+
+// clang -O2 -Werror -c droppacket_unsafe.c -o droppacket_unsafe_jit.o
+//
+// For bpf code: clang -target bpf -O2 -Werror -c droppacket_unsafe.c -o droppacket_unsafe.o
```
Comment (dthaler) [inline tests/sample/empty.c:6]:
```suggestion
// clang -O2 -Werror -c empty.c -o empty_jit.o
//
// For bpf code: clang -target bpf -O2 -Werror -c empty.c -o empty.o
```

File: tests/sample/empty.c [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+#include "bpf_endian.h"
+#include "bpf_helpers.h"
+#include "net/ip.h"
+#include "net/udp.h"
+
+SEC("maps")
+struct bpf_map_def port_map = {
+    .type = BPF_MAP_TYPE_ARRAY, .key_size = sizeof(uint32_t), .value_size = sizeof(uint64_t), .max_entries = 1};
+
+// EBPF program with no segments
```
Comment (dthaler) [inline tests/sample/empty.c:28]:
```suggestion
// No code sections.
```
It's not an eBPF "program" per se.  program_count will be 0.

File: tools/bpf2c/bpf_code_generator.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply tools/bpf2c/bpf_code_generator.cpp:947]: Blocking it in bpf2c.cpp

**Review by saxena-anurag — APPROVED** [review]

**Review by dthaler — APPROVED** [review]

**Pull request merged**